calico.go

```diff
@@ -44,80 +44,91 @@ func init() {
 	// since namespace ops (unshare, setns) are done for a single thread, we
 	// must ensure that the goroutine does not jump from OS thread to thread
 	runtime.LockOSThread()
-
-	nodename, _ = os.Hostname()
-}
-
-func updateNodename(conf NetConf, logger *log.Entry) {
-	if conf.Hostname != "" {
-		nodename = conf.Hostname
-		logger.Warn("Configuration option 'hostname' is deprecated, use 'nodename' instead.")
-	}
-	if conf.Nodename != "" {
-		nodename = conf.Nodename
-	}
 }
 
 func cmdAdd(args *skel.CmdArgs) error {
 	// Unmarshal the network config, and perform validation
-	conf := NetConf{}
+	conf := types.NetConf{}
 	if err := json.Unmarshal(args.StdinData, &conf); err != nil {
 		return fmt.Errorf("failed to load netconf: %v", err)
 	}
 
-	cniVersion := conf.CNIVersion
+	utils.ConfigureLogging(conf.LogLevel)
 
-	ConfigureLogging(conf.LogLevel)
+	// Allow the nodename to be overridden by the network config
+	nodename = utils.DetermineNodename(conf)
 
-	workload, orchestrator, err := GetIdentifiers(args)
+	// Extract WEP identifiers such as pod name, pod namespace (for k8s), containerID, workload name.
+	epIDs, err := utils.GetIdentifiers(args, nodename)
 	if err != nil {
 		return err
 	}
 
-	logger := CreateContextLogger(workload)
-
-	// Allow the nodename to be overridden by the network config
-	updateNodename(conf, logger)
-
-	logger.WithFields(log.Fields{
-		"Orchestrator": orchestrator,
-		"Node":         nodename,
-		"Workload":     workload,
-		"ContainerID":  args.ContainerID,
-	}).Info("Extracted identifiers")
+	logger := &log.Entry{}
+	if epIDs.Orchestrator == "k8s" {
+		logger = log.WithFields(log.Fields{
+			"WorkloadEndpoint": epIDs.Workload,
+			"ContainerID":      epIDs.ContainerID,
+			"Pod":              epIDs.Pod,
+			"Namespace":        epIDs.Namespace,
+		})
+	} else {
+		logger = log.WithFields(log.Fields{
+			"WorkloadEndpoint": epIDs.Workload,
+			"ContainerID":      epIDs.ContainerID,
+		})
+	}
+	logger.Info("Extracted identifiers")
+	logger.WithField("NetConfg", conf).Info("Loaded CNI NetConf")
 
-	logger.WithFields(log.Fields{"NetConfg": conf}).Info("Loaded CNI NetConf")
-	calicoClient, err := CreateClient(conf)
+	calicoClient, err := utils.CreateClient(conf)
 	if err != nil {
 		return err
 	}
 
-	// Always check if there's an existing endpoint.
-	endpoints, err := calicoClient.WorkloadEndpoints().List(api.WorkloadEndpointMetadata{
-		Node:         nodename,
-		Orchestrator: orchestrator,
-		Workload:     workload})
+	ctx := context.Background()
+
+	// Check if there's an existing endpoint.
+	endpoints, err := calicoClient.WorkloadEndpoints().List(ctx, options.ListOptions{Name: epIDs.Workload, Namespace: epIDs.Namespace, Prefix: true})
 	if err != nil {
 		return err
 	}
 
-	logger.Debugf("Retrieved endpoints: %v", endpoints)
-
 	var endpoint *api.WorkloadEndpoint
-	if len(endpoints.Items) == 1 {
-		endpoint = &endpoints.Items[0]
+
+	if len(endpoints.Items) >= 1 {
```

Comment:
>= 1 ;  i think >0 is easier to grok at a glance.

Reply:
also, this check is redundant.  if there are 0 entries in the Items, then the for loop won't do anything.

Reply:
agree with `>0` part, but I'd still like to keep that if statement because it's more explicit